src/porepy/numerics/fv/tpfa.py

```diff
@@ -266,3 +270,462 @@ def discretize(self, sd: pp.Grid, data: dict) -> None:
+                raise NotImplementedError
+            else:
+                mat_loc = grid_property_getter(g)
+            blocks.append(mat_loc)
+
+        block_matrix = pp.matrix_operations.optimized_compressed_storage(
+            sps.block_diag(blocks)
+        )
+        return block_matrix
+
+    def half_face_map(
+        self,
+        subdomains: list[pp.Grid],
+        from_entity: Literal["cells", "faces", "half_faces"] = "half_faces",
+        to_entity: Literal["cells", "faces", "half_faces"] = "half_faces",
+        dimensions: tuple[int, int] = (1, 1),
+        with_sign: bool = False,
+    ) -> sps.spmatrix:
+        """Mapping between half-faces and cells or faces.
+
+        Parameters:
+            subdomains: List of grids.
+            from_entity: Entity to map from.
+            to_entity: Entity to map to.
+            dimensions: Dimensions of the to and from entities. EK: Verify order
+            with_sign: Whether to include sign in the mapping.
+
+        Returns:
+            spmatrix ``(num_{from_entity} * dimensions[0],
+                        num_{to_entity} * dimensions[1])``:
+                Mapping matrix.
+
+        """
+
+        def get_matrix(g: pp.Grid) -> sps.csr_matrix:
+            fi, ci, sgn = sps.find(g.cell_faces)
+
+            indices = []
+            sizes = []
+            for name in [to_entity, from_entity]:
+                if name == "cells":
+                    indices.append(ci)
+                    sizes.append(g.num_cells)
+                elif name == "faces":
+                    indices.append(fi)
+                    sizes.append(g.num_faces)
+                elif name == "half_faces":
+                    indices.append(np.arange(fi.size))
+                    sizes.append(fi.size)
+                else:
+                    raise ValueError(f"Unknown entity {name}.")
+
+            # If the dimensions of the from and to entities are unequal, we need to
+            # repeat the indices of the lowest dimension to match the highest dimension.
+            # Find the ratio between the dimensions to know how many times to repeat.
+            # Ceil is used to treat the lower of the two dimensions as 1.
+            repeat_row_inds = np.ceil(dimensions[1] / dimensions[0]).astype(int)
+            repeat_col_inds = np.ceil(dimensions[0] / dimensions[1]).astype(int)
+
+            # Check that the dimensions are compatible. One of the dimensions must be a
+            # multiple of the other.
+            assert (
+                dimensions[0] % dimensions[1] == 0 or dimensions[1] % dimensions[0] == 0
+            )
+            # Expand the indices, with the following logic (for rows): First repeat the
+            # row indices to match the size of columns (the from entity). Then expand
+            # these indices to account for the dimenison of the to entity
+            # (dimensions[0]). The latter assigns rows[indices 0:dimensions[0]] to the
```

Comment:
Is there something a bit off with the indexing of rows?

Reply:
I don't understand the question.

Reply:
I see. I just pushed a minor update, should be better now.